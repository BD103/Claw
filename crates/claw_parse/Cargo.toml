[package]
name = "claw_parse"
version = "0.1.0-dev"
authors = ["BD103"]
edition = "2021"
description = "Programming language that compiles into Scratch blocks (SB3)."
readme = "README.md"
# homepage = "https://bd103.github.io/Claw"
repository = "https://github.com/BD103/Claw"
license = "MIT OR Apache-2.0"
publish = false

[dependencies]
ariadne = "0.1.5"
chumsky = "0.8.0"

# Required for snapshot testing
serde = { version = "1.0.145", optional = true, features = ["derive"] }

[dev-dependencies]
insta = { version = "1.20.0", features = ["glob", "ron"] }

[[test]]
name = "snap"
path = "tests/snap.rs"
required-features = ["serde"]
