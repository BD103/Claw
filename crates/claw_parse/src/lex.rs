use claw_middle::Span;
use logos::Logos;

pub fn lex<'source>(source: &'source str) -> Tokenizer<'source> {
    Tokenizer::new(source)
}

pub struct Tokenizer<'source> {
    inner: logos::Lexer<'source, TokenKind>,
}

impl<'source> Tokenizer<'source> {
    fn new(source: &'source str) -> Self {
        Tokenizer {
            inner: TokenKind::lexer(source),
        }
    }
}

impl<'source> Iterator for Tokenizer<'source> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        Some(Token {
            kind: self.inner.next()?,
            span: self.inner.span(),
        })
    }
}

pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

#[derive(Logos, Debug, PartialEq)]
#[non_exhaustive]
pub enum TokenKind {
    #[error]
    Error,
}
