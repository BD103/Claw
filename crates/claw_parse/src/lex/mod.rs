mod token;

pub use self::token::{Token, TokenKind, token};

pub fn lex<'a>(source: &'a str) -> Tokenizer<'a> {
    Tokenizer::new(source)
}

pub struct Tokenizer<'a> {
    cursor: usize,
    source: &'a str,
}

impl<'a> Tokenizer<'a> {
    fn new(source: &'a str) -> Self {
        Tokenizer {
            cursor: 0,
            source,
        }
    }
}

impl Iterator for Tokenizer<'_> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        if self.cursor > self.source.len() - 1 {
            return Some(Token {
                kind: token!(EOF),
                span: 0..0,
            });
        }

        todo!()
    }
}
