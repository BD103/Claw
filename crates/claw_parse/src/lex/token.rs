use std::ops::Range;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Token {
    pub kind: TokenKind,
    pub span: Range<usize>,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub enum TokenKind {
    // Values
    /// Either an identifier or keyword.
    ///
    /// `fn`, `looks`, etc.
    Ident(String),
    /// A string.
    ///
    /// `"Some text"`
    Text(String),
    /// A number, optional decimal and sign.
    ///
    /// `64`, `-2`, `3.1415`
    Number(String),
    /// True or false.
    ///
    /// `true` or `false`
    Boolean(bool),

    // Delimiters
    /// `(`
    ParenOpen,
    /// `)`
    ParenClose,
    /// `{`
    BracketOpen,
    /// `}`
    BracketClose,

    // Separators
    /// `::`
    DoubleColon,
    /// `:`
    Colon,
    /// `;`
    Semi,
    /// `,`
    Comma,

    // Prefixes
    /// `@`
    AtSign,
    /// `$`
    EnumSign,

    // Comparison
    /// `==`
    Eq,
    /// `!=`
    NotEq,
    /// `&&`
    And,
    /// `||`
    Or,
    /// `>`
    Gt,
    /// `>=`
    Ge,
    /// `<`
    Lt,
    /// `<=`
    Le,
}
