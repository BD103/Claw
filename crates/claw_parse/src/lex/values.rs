use super::TokenKind;
use crate::error::LexError;
use chumsky::prelude::*;

pub fn create_ident() -> impl Parser<char, TokenKind, Error = LexError> {
    text::ident().map(TokenKind::Ident)
}

pub fn create_text() -> impl Parser<char, TokenKind, Error = LexError> {
    just('"')
        .ignore_then(filter(|c| *c != '"').repeated())
        .then_ignore(just('"'))
        .collect::<String>()
        .map(TokenKind::Text)
}

pub fn create_number() -> impl Parser<char, TokenKind, Error = LexError> {
    let num = text::int(10)
        .chain::<char, _, _>(just('.').chain(text::digits(10)).or_not().flatten())
        .collect::<String>();

    // Supports negatives
    choice((
        just('-').then(num.clone()).map(|(_, mut x)| {
            x.insert(0, '-');
            TokenKind::Number(x)
        }),
        num.map(TokenKind::Number),
    ))
}

pub fn create_bool() -> impl Parser<char, TokenKind, Error = LexError> {
    text::keyword("true")
        .ignored()
        .to(TokenKind::Boolean(true))
        .or(text::keyword("false")
            .ignored()
            .to(TokenKind::Boolean(false)))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ident() {
        let input = [
            "hi",
            "hiThere",
            "hi_there",
            "_hiThErE",
            "h1Th3r3",
        ];
        // Invalid idents
        let input_err = [
            "5ithere",
            "-hello",
            "ğŸ¦€",
        ];

        let parser = create_ident();

        input.into_iter()
            .for_each(|x| {
                let output = parser.parse(x);
                assert_eq!(output, Ok(TokenKind::Ident(x.into())));
            });
        
        input_err.into_iter()
            .for_each(|x| {
                let output = parser.parse(x);
                assert!(output.is_err());
            });
        
    }
}
