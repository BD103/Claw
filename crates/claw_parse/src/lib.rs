#![doc = include_str!("../README.md")]

pub mod ast;
pub mod error;
pub mod lex;
pub mod parse;

use ariadne::Report;
use chumsky::Parser;

pub fn parse(script: &str) -> Result<(), Box<Report>> {
    let lexer = lex::create_lexer();
    let tokens = lexer.parse(script);

    let tokens = match tokens {
        Ok(tokens) => tokens,
        Err(errors) => {
            return Err(Box::new(error::generate_token_report(errors)));
        }
    };

    let parser = parse::create_parser();
    let ast = parser.parse(tokens);

    match ast {
        Ok(ast) => {
            dbg!(ast);
        },
        Err(errors) => {
            dbg!(errors);
        },
    };

    Ok(())
}
