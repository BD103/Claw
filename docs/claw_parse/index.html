<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="`claw_parse`"><meta name="keywords" content="rust, rustlang, rust-lang, claw_parse"><title>claw_parse - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../dark.css" disabled><link rel="stylesheet" type="text/css" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script defer src="../crates.js"></script><script defer src="../main.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../claw_parse/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location"></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../claw_parse/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location"><a href="#">Crate claw_parse</a></h2><div class="sidebar-elems"><div class="block"><ul><li class="version">Version 0.1.0-dev</li><li><a id="all-types" href="all.html">All Items</a></li></ul></div><section><div class="block"><ul><li><a href="#reexports">Re-exports</a></li><li><a href="#modules">Modules</a></li><li><a href="#functions">Functions</a></li><li><a href="#types">Type Definitions</a></li></ul></div></section></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../claw_parse/index.html"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><button type="button">?</button></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div></div></form></nav></div><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn"><span class="in-band">Crate <a class="mod" href="#">claw_parse</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../src/claw_parse/lib.rs.html#1-25">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><h2 id="claw_parse"><a href="#claw_parse"><code>claw_parse</code></a></h2>
<p>This crate is part of the Claw programming language. You can see the project README <a href="https://github.com/BD103/Claw">here</a>.</p>
</div></details><h2 id="reexports" class="small-section-header"><a href="#reexports">Re-exports</a></h2><div class="item-table"><div class="item-row"><div class="item-left import-item" id="reexport.AST"><code>pub use crate::ast::<a class="type" href="ast/type.AST.html" title="type claw_parse::ast::AST">AST</a>;</code></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left import-item" id="reexport.create_parser"><code>pub use crate::parser::<a class="fn" href="parser/fn.create_parser.html" title="fn claw_parse::parser::create_parser">create_parser</a>;</code></div><div class="item-right docblock-short"></div></div></div><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="mod" href="ast/index.html" title="claw_parse::ast mod">ast</a></div><div class="item-right docblock-short"><p>Contains the abstract syntax tree definitions.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="parser/index.html" title="claw_parse::parser mod">parser</a></div><div class="item-right docblock-short"><p>Contains the logic of the actual parser, written using [<code>chumsky</code>].</p>
</div></div></div><h2 id="functions" class="small-section-header"><a href="#functions">Functions</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.build_report.html" title="claw_parse::build_report fn">build_report</a></div><div class="item-right docblock-short"><p>Transforms a result to use [<code>ariadne</code>]’s [<code>Report</code>].</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.get_source.html" title="claw_parse::get_source fn">get_source</a></div><div class="item-right docblock-short"><p>Returns the [<code>Source</code>] of a string reference.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.parse.html" title="claw_parse::parse fn">parse</a></div><div class="item-right docblock-short"><p>Parses a <a href="https://doc.rust-lang.org/1.64.0/std/primitive.str.html" title="str"><code>str</code></a> into <a href="ast/type.AST.html" title="AST"><code>AST</code></a>.</p>
</div></div></div><h2 id="types" class="small-section-header"><a href="#types">Type Definitions</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="type" href="type.ParseError.html" title="claw_parse::ParseError type">ParseError</a></div><div class="item-right docblock-short"><p>Error type returned by the parsers.</p>
</div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="claw_parse" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.64.0 (a55dd71d5 2022-09-19)" ></div></body></html>