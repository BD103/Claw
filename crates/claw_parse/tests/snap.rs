// If this isn't running when you execute test, make sure to enable the "snap" feature. This will be
// skipped if disabled. https://doc.rust-lang.org/stable/cargo/reference/cargo-targets.html#the-required-features-field

use insta::{assert_ron_snapshot, glob};
use std::fs::read_to_string;

#[test]
fn lexer() {
    use chumsky::Parser;
    use claw_parse::lex::*;

    let lexer = create_lexer();

    glob!("samples/*.claw", |path| {
        let res = lexer.parse(read_to_string(path).unwrap());

        assert!(res.is_ok());
        assert_ron_snapshot!(res.unwrap());
    });
}
