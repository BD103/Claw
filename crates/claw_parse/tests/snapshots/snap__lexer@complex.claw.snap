---
source: crates/claw_parse/tests/snap.rs
expression: res.unwrap()
input_file: crates/claw_parse/tests/samples/complex.claw
---
[
  Token(
    kind: Ident("declare"),
    span: Range(
      start: 0,
      end: 7,
    ),
  ),
  Token(
    kind: Ident("Var"),
    span: Range(
      start: 8,
      end: 11,
    ),
  ),
  Token(
    kind: BracketOpen,
    span: Range(
      start: 12,
      end: 13,
    ),
  ),
  Token(
    kind: Ident("Score"),
    span: Range(
      start: 18,
      end: 23,
    ),
  ),
  Token(
    kind: Comma,
    span: Range(
      start: 23,
      end: 24,
    ),
  ),
  Token(
    kind: Ident("HighScore"),
    span: Range(
      start: 29,
      end: 38,
    ),
  ),
  Token(
    kind: Comma,
    span: Range(
      start: 38,
      end: 39,
    ),
  ),
  Token(
    kind: BracketClose,
    span: Range(
      start: 64,
      end: 65,
    ),
  ),
  Token(
    kind: AtSign,
    span: Range(
      start: 67,
      end: 68,
    ),
  ),
  Token(
    kind: Ident("on"),
    span: Range(
      start: 68,
      end: 70,
    ),
  ),
  Token(
    kind: ParenOpen,
    span: Range(
      start: 70,
      end: 71,
    ),
  ),
  Token(
    kind: Ident("start"),
    span: Range(
      start: 71,
      end: 76,
    ),
  ),
  Token(
    kind: ParenClose,
    span: Range(
      start: 76,
      end: 77,
    ),
  ),
  Token(
    kind: Ident("fn"),
    span: Range(
      start: 78,
      end: 80,
    ),
  ),
  Token(
    kind: Ident("start"),
    span: Range(
      start: 81,
      end: 86,
    ),
  ),
  Token(
    kind: ParenOpen,
    span: Range(
      start: 86,
      end: 87,
    ),
  ),
  Token(
    kind: ParenClose,
    span: Range(
      start: 87,
      end: 88,
    ),
  ),
  Token(
    kind: BracketOpen,
    span: Range(
      start: 89,
      end: 90,
    ),
  ),
  Token(
    kind: Ident("looks"),
    span: Range(
      start: 95,
      end: 100,
    ),
  ),
  Token(
    kind: DoubleColon,
    span: Range(
      start: 100,
      end: 102,
    ),
  ),
  Token(
    kind: Ident("switch_backdrop"),
    span: Range(
      start: 102,
      end: 117,
    ),
  ),
  Token(
    kind: ParenOpen,
    span: Range(
      start: 117,
      end: 118,
    ),
  ),
  Token(
    kind: EnumSign,
    span: Range(
      start: 118,
      end: 119,
    ),
  ),
  Token(
    kind: Ident("Backdrop"),
    span: Range(
      start: 119,
      end: 127,
    ),
  ),
  Token(
    kind: DoubleColon,
    span: Range(
      start: 127,
      end: 129,
    ),
  ),
  Token(
    kind: Ident("Blue"),
    span: Range(
      start: 129,
      end: 133,
    ),
  ),
  Token(
    kind: ParenClose,
    span: Range(
      start: 133,
      end: 134,
    ),
  ),
  Token(
    kind: Semi,
    span: Range(
      start: 134,
      end: 135,
    ),
  ),
  Token(
    kind: BracketClose,
    span: Range(
      start: 136,
      end: 137,
    ),
  ),
  Token(
    kind: Ident("sprite"),
    span: Range(
      start: 139,
      end: 145,
    ),
  ),
  Token(
    kind: Ident("Scratch"),
    span: Range(
      start: 146,
      end: 153,
    ),
  ),
  Token(
    kind: BracketOpen,
    span: Range(
      start: 154,
      end: 155,
    ),
  ),
  Token(
    kind: Ident("declare"),
    span: Range(
      start: 193,
      end: 200,
    ),
  ),
  Token(
    kind: Ident("Var"),
    span: Range(
      start: 201,
      end: 204,
    ),
  ),
  Token(
    kind: BracketOpen,
    span: Range(
      start: 205,
      end: 206,
    ),
  ),
  Token(
    kind: Ident("AccY"),
    span: Range(
      start: 207,
      end: 211,
    ),
  ),
  Token(
    kind: BracketClose,
    span: Range(
      start: 212,
      end: 213,
    ),
  ),
  Token(
    kind: AtSign,
    span: Range(
      start: 219,
      end: 220,
    ),
  ),
  Token(
    kind: Ident("on"),
    span: Range(
      start: 220,
      end: 222,
    ),
  ),
  Token(
    kind: ParenOpen,
    span: Range(
      start: 222,
      end: 223,
    ),
  ),
  Token(
    kind: Ident("start"),
    span: Range(
      start: 223,
      end: 228,
    ),
  ),
  Token(
    kind: ParenClose,
    span: Range(
      start: 228,
      end: 229,
    ),
  ),
  Token(
    kind: Ident("fn"),
    span: Range(
      start: 234,
      end: 236,
    ),
  ),
  Token(
    kind: Ident("start"),
    span: Range(
      start: 237,
      end: 242,
    ),
  ),
  Token(
    kind: ParenOpen,
    span: Range(
      start: 242,
      end: 243,
    ),
  ),
  Token(
    kind: ParenClose,
    span: Range(
      start: 243,
      end: 244,
    ),
  ),
  Token(
    kind: BracketOpen,
    span: Range(
      start: 245,
      end: 246,
    ),
  ),
  Token(
    kind: Ident("use"),
    span: Range(
      start: 255,
      end: 258,
    ),
  ),
  Token(
    kind: Ident("sensing"),
    span: Range(
      start: 259,
      end: 266,
    ),
  ),
  Token(
    kind: DoubleColon,
    span: Range(
      start: 266,
      end: 268,
    ),
  ),
  Token(
    kind: Ident("is_key_pressed"),
    span: Range(
      start: 268,
      end: 282,
    ),
  ),
  Token(
    kind: Semi,
    span: Range(
      start: 282,
      end: 283,
    ),
  ),
  Token(
    kind: Ident("motion"),
    span: Range(
      start: 293,
      end: 299,
    ),
  ),
  Token(
    kind: DoubleColon,
    span: Range(
      start: 299,
      end: 301,
    ),
  ),
  Token(
    kind: Ident("goto"),
    span: Range(
      start: 301,
      end: 305,
    ),
  ),
  Token(
    kind: ParenOpen,
    span: Range(
      start: 305,
      end: 306,
    ),
  ),
  Token(
    kind: Number("0"),
    span: Range(
      start: 306,
      end: 307,
    ),
  ),
  Token(
    kind: Comma,
    span: Range(
      start: 307,
      end: 308,
    ),
  ),
  Token(
    kind: Number("0"),
    span: Range(
      start: 309,
      end: 310,
    ),
  ),
  Token(
    kind: ParenClose,
    span: Range(
      start: 310,
      end: 311,
    ),
  ),
  Token(
    kind: Semi,
    span: Range(
      start: 311,
      end: 312,
    ),
  ),
  Token(
    kind: Ident("forever"),
    span: Range(
      start: 322,
      end: 329,
    ),
  ),
  Token(
    kind: BracketOpen,
    span: Range(
      start: 330,
      end: 331,
    ),
  ),
  Token(
    kind: Ident("if"),
    span: Range(
      start: 344,
      end: 346,
    ),
  ),
  Token(
    kind: Ident("is_key_pressed"),
    span: Range(
      start: 347,
      end: 361,
    ),
  ),
  Token(
    kind: ParenOpen,
    span: Range(
      start: 361,
      end: 362,
    ),
  ),
  Token(
    kind: EnumSign,
    span: Range(
      start: 362,
      end: 363,
    ),
  ),
  Token(
    kind: Ident("Key"),
    span: Range(
      start: 363,
      end: 366,
    ),
  ),
  Token(
    kind: DoubleColon,
    span: Range(
      start: 366,
      end: 368,
    ),
  ),
  Token(
    kind: Ident("Up"),
    span: Range(
      start: 368,
      end: 370,
    ),
  ),
  Token(
    kind: ParenClose,
    span: Range(
      start: 370,
      end: 371,
    ),
  ),
  Token(
    kind: BracketOpen,
    span: Range(
      start: 372,
      end: 373,
    ),
  ),
  Token(
    kind: Ident("self"),
    span: Range(
      start: 390,
      end: 394,
    ),
  ),
  Token(
    kind: DoubleColon,
    span: Range(
      start: 394,
      end: 396,
    ),
  ),
  Token(
    kind: Ident("acc_y"),
    span: Range(
      start: 396,
      end: 401,
    ),
  ),
  Token(
    kind: ParenOpen,
    span: Range(
      start: 401,
      end: 402,
    ),
  ),
  Token(
    kind: Number("0.1"),
    span: Range(
      start: 402,
      end: 405,
    ),
  ),
  Token(
    kind: ParenClose,
    span: Range(
      start: 405,
      end: 406,
    ),
  ),
  Token(
    kind: Semi,
    span: Range(
      start: 406,
      end: 407,
    ),
  ),
  Token(
    kind: BracketClose,
    span: Range(
      start: 420,
      end: 421,
    ),
  ),
  Token(
    kind: Ident("else"),
    span: Range(
      start: 422,
      end: 426,
    ),
  ),
  Token(
    kind: Ident("if"),
    span: Range(
      start: 427,
      end: 429,
    ),
  ),
  Token(
    kind: Ident("is_key_pressed"),
    span: Range(
      start: 430,
      end: 444,
    ),
  ),
  Token(
    kind: ParenOpen,
    span: Range(
      start: 444,
      end: 445,
    ),
  ),
  Token(
    kind: EnumSign,
    span: Range(
      start: 445,
      end: 446,
    ),
  ),
  Token(
    kind: Ident("Key"),
    span: Range(
      start: 446,
      end: 449,
    ),
  ),
  Token(
    kind: DoubleColon,
    span: Range(
      start: 449,
      end: 451,
    ),
  ),
  Token(
    kind: Ident("Down"),
    span: Range(
      start: 451,
      end: 455,
    ),
  ),
  Token(
    kind: ParenClose,
    span: Range(
      start: 455,
      end: 456,
    ),
  ),
  Token(
    kind: BracketOpen,
    span: Range(
      start: 457,
      end: 458,
    ),
  ),
  Token(
    kind: Ident("data"),
    span: Range(
      start: 513,
      end: 517,
    ),
  ),
  Token(
    kind: DoubleColon,
    span: Range(
      start: 517,
      end: 519,
    ),
  ),
  Token(
    kind: Ident("set_var"),
    span: Range(
      start: 519,
      end: 526,
    ),
  ),
  Token(
    kind: ParenOpen,
    span: Range(
      start: 526,
      end: 527,
    ),
  ),
  Token(
    kind: EnumSign,
    span: Range(
      start: 527,
      end: 528,
    ),
  ),
  Token(
    kind: Ident("Var"),
    span: Range(
      start: 528,
      end: 531,
    ),
  ),
  Token(
    kind: DoubleColon,
    span: Range(
      start: 531,
      end: 533,
    ),
  ),
  Token(
    kind: Ident("AccY"),
    span: Range(
      start: 533,
      end: 537,
    ),
  ),
  Token(
    kind: Comma,
    span: Range(
      start: 537,
      end: 538,
    ),
  ),
  Token(
    kind: Number("0"),
    span: Range(
      start: 539,
      end: 540,
    ),
  ),
  Token(
    kind: ParenClose,
    span: Range(
      start: 540,
      end: 541,
    ),
  ),
  Token(
    kind: Semi,
    span: Range(
      start: 541,
      end: 542,
    ),
  ),
  Token(
    kind: BracketClose,
    span: Range(
      start: 555,
      end: 556,
    ),
  ),
  Token(
    kind: Ident("self"),
    span: Range(
      start: 570,
      end: 574,
    ),
  ),
  Token(
    kind: DoubleColon,
    span: Range(
      start: 574,
      end: 576,
    ),
  ),
  Token(
    kind: Ident("apply_acc"),
    span: Range(
      start: 576,
      end: 585,
    ),
  ),
  Token(
    kind: ParenOpen,
    span: Range(
      start: 585,
      end: 586,
    ),
  ),
  Token(
    kind: ParenClose,
    span: Range(
      start: 586,
      end: 587,
    ),
  ),
  Token(
    kind: Semi,
    span: Range(
      start: 587,
      end: 588,
    ),
  ),
  Token(
    kind: BracketClose,
    span: Range(
      start: 597,
      end: 598,
    ),
  ),
  Token(
    kind: BracketClose,
    span: Range(
      start: 603,
      end: 604,
    ),
  ),
  Token(
    kind: Ident("fn"),
    span: Range(
      start: 610,
      end: 612,
    ),
  ),
  Token(
    kind: Ident("acc_y"),
    span: Range(
      start: 613,
      end: 618,
    ),
  ),
  Token(
    kind: ParenOpen,
    span: Range(
      start: 618,
      end: 619,
    ),
  ),
  Token(
    kind: Ident("n"),
    span: Range(
      start: 619,
      end: 620,
    ),
  ),
  Token(
    kind: Colon,
    span: Range(
      start: 620,
      end: 621,
    ),
  ),
  Token(
    kind: Ident("Number"),
    span: Range(
      start: 622,
      end: 628,
    ),
  ),
  Token(
    kind: ParenClose,
    span: Range(
      start: 628,
      end: 629,
    ),
  ),
  Token(
    kind: BracketOpen,
    span: Range(
      start: 630,
      end: 631,
    ),
  ),
  Token(
    kind: Ident("if"),
    span: Range(
      start: 687,
      end: 689,
    ),
  ),
  Token(
    kind: Ident("ops"),
    span: Range(
      start: 690,
      end: 693,
    ),
  ),
  Token(
    kind: DoubleColon,
    span: Range(
      start: 693,
      end: 695,
    ),
  ),
  Token(
    kind: Ident("add"),
    span: Range(
      start: 695,
      end: 698,
    ),
  ),
  Token(
    kind: ParenOpen,
    span: Range(
      start: 698,
      end: 699,
    ),
  ),
  Token(
    kind: EnumSign,
    span: Range(
      start: 699,
      end: 700,
    ),
  ),
  Token(
    kind: Ident("Var"),
    span: Range(
      start: 700,
      end: 703,
    ),
  ),
  Token(
    kind: DoubleColon,
    span: Range(
      start: 703,
      end: 705,
    ),
  ),
  Token(
    kind: Ident("AccY"),
    span: Range(
      start: 705,
      end: 709,
    ),
  ),
  Token(
    kind: Comma,
    span: Range(
      start: 709,
      end: 710,
    ),
  ),
  Token(
    kind: Ident("n"),
    span: Range(
      start: 711,
      end: 712,
    ),
  ),
  Token(
    kind: ParenClose,
    span: Range(
      start: 712,
      end: 713,
    ),
  ),
]
